### DISCUSSION TITLE
Adding Swagger to Genie

### DISCUSSION THREAD

**Discussion by jiachengzhang1** [GenieFramework/Genie.jl#509] [Ideas]
I'm seeing  some attractions around [SwagUI](https://github.com/jiachengzhang1/SwagUI.jl), thanks some of you for checking out! Now I plan to optimize it a bit, just wonder if it there's a need to built it into the Genie.jl so that users can have it out-of-the-box (similar to fastapi), I can work on this, or you'd prefer users to add it separately. What are your thoughts?

**Comment by essenciary** [comment]
Thank you @jiachengzhang1 - great question :) 

The strategy for the upcoming Genie v5 is to move towards
a) unbundling of features - this is the case with non-core features, especially if they have extra dependencies, in order to make Genie core small and agile, reduce pre-compilation time, etc. Modules that will be unbundled in v5 include `Sessions` and `Cache`. This will allow us to develop `Sessions` and `Cache` separately and grow them for example by adding a multitude of backends for both (files, relational db, in-memory storage dbs, etc).
b) make it very easy to install needed dependencies by providing wizards and potentially a Genie plugin manager. Ex with the wizards, when developing a full stack MVC app it will suggest installing the `Sessions` and `Cache` plugins. A future plugin manager would allow searching for plugins by feature, across all available packages. 

With SwagUI there are no extra dependencies - but I would still keep it as a plugin and we could integrate it into a Genie API wizard. In general, keeping these features in distinct plugins will allow interesting benefits, like: 
* the maintainers don't have to deal with Genie's internals
* separate release cycle (eg these can move to breaking v2, v3, etc while offering support for various Genie versions -- no need to wait for breaking Genie releases)

What do you think?

**Reply by jiachengzhang1** [reply]
I like this idea, it does make the development smoother, looking forward to the growing plugins. In the meantime, I'll keep improving SwagUI and add it as a plugin.

**Comment by eliteuser26** [comment]
I would be interested in having SwagUI as a plugin in Genie. I do use this feature in Python with FastAPI for testing my web pages. It is useful in finding errors when using post payload data.

**Comment by essenciary** [comment]
That's already available. See https://github.com/GenieFramework/PkgVizBoard/blob/master/routes.jl
